### PR TITLE
Fix multi gpu LM issues and add hdf5 LM dataset dump

### LINKED ISSUE #1075: multi gpu lm training throws CUDA out of memory exception
There was an out of memory exception thrown when I use lm.yaml and lm_large.yaml in librispeech egs with multi gpu=4.
```
Exception in main training loop: CUDA out of memory. Tried to allocate 19.62 MiB (GPU 2; 22.38 GiB total capacity; 21.01 GiB already allocated; 10.06 MiB free; 634.29 MiB cached)                                   
Traceback (most recent call last):
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/trainer.py", line 316, in run                                                                                              
    update()
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/updaters/standard_updater.py", line 175, in update                                                                         
    self.update_core()
  File "/mnt/cfs1/asr/users/fanlu/espnet/espnet/lm/pytorch_backend/lm.py", line 277, in update_core
    loss.backward()  # Backprop
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/tensor.py", line 102, in backward                                                                                                     
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/autograd/__init__.py", line 90, in backward                                                                                           
    allow_unreachable=True)  # allow_unreachable flag
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "/mnt/cfs1/asr/users/fanlu/espnet/espnet/bin/lm_train.py", line 156, in <module>
    main(sys.argv[1:])
  File "/mnt/cfs1/asr/users/fanlu/espnet/espnet/bin/lm_train.py", line 150, in main
    train(args)
  File "/mnt/cfs1/asr/users/fanlu/espnet/espnet/lm/pytorch_backend/lm.py", line 416, in train
    trainer.run()
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/trainer.py", line 349, in run                                                                                              
    six.reraise(*exc_info)
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/six.py", line 693, in reraise
    raise value
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/trainer.py", line 316, in run                                                                                              
    update()
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/updaters/standard_updater.py", line 175, in update                                                                         
    self.update_core()
  File "/mnt/cfs1/asr/users/fanlu/espnet/espnet/lm/pytorch_backend/lm.py", line 277, in update_core
    loss.backward()  # Backprop
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/tensor.py", line 102, in backward                                                                                                     
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/mnt/cfs1/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/autograd/__init__.py", line 90, in backward                                                                                           
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: CUDA out of memory. Tried to allocate 19.62 MiB (GPU 2; 22.38 GiB total capacity; 21.01 GiB already allocated; 10.06 MiB free; 634.29 MiB cached)                                                      
# Accounting: time=746 threads=1
# Finished at Sun Aug 11 14:58:25 CST 2019 with status 1
```

### PR DESCRIPTION
**Pull request by ShigekiKarita**
This PR aims to fix #1075 . I tested this by AN4 and TEDLIUM3. 

I found @kan-bayashi's comment was correct. The memory leak appeared when single GPU performs the frame reduction of loss values from multi GPUs. So I added the `ReduceFramewiseLoss` module to do it in multi GPUs.

In addition, I added HDF5 dump support for LM dataset because some large corpora, e.g., TEDLIUM3, results in 3GB of text dataset and it took 3 mins for preprocessing. After HDF5 dump performed, it only took 30 secs to load that. It really helps me testing this problem, and running rapid try-and-errors in the large LM experiment. (PS. I also tested pickle but it seems to perform very slow and save large bin.)

### CODE REVIEW

File: espnet/bin/lm_train.py [hunk 1]
```diff
@@ -85,6 +85,8 @@ def get_parser():
                         help='dropout probability')
     parser.add_argument('--maxlen', type=int, default=40,
                         help='Batch size is reduced if the input sequence > ML')
+    parser.add_argument('--dump-dataset', type=str, default=None,
```
Comment (kan-bayashi) [inline espnet/bin/lm_train.py]: I think it is better to rename or change to bool type to make it clear.

Comment (ShigekiKarita) [reply espnet/bin/lm_train.py]: Maybe `--dump-hdf5-path` is better?

Comment (kan-bayashi) [reply espnet/bin/lm_train.py]: Sounds good :)

**Comment by ShigekiKarita** [comment]
Let me update some docstrings...

**Comment by ShigekiKarita** [comment]
done

**Comment by fanlu** [comment]
Cool, I have tested this pr, it worked.
There is a problem in asr training with a big dataset such as 6000 hours audio data also.
I suspect these are the same problem.
```
 ^[[J     total [..................................................]  0.34%                                                        
 this epoch [########..........................................] 17.05%                                                            
      37300 iter, 0 epoch / 50 epochs                                                                                              
    0.76965 iters/sec. Estimated time to finish: 163 days, 21:38:20.292327.                                                        
 ^[[4A^[[J0           37401       18.2174     18.4124        18.                                                                    1338                                                                                  0.                                           685313                                                                64476.4                                                     
 ^[[J0           37500       15.5689     17.2413        14.                                                                         8522                                                                                  0.                                           72104                                                                 64593.9                                                     
 ^[[J     total [..................................................]  0.34%                                                        
 this epoch [########..........................................] 17.14%                                                            
      37500 iter, 0 epoch / 50 epochs                                                                                              
    0.77792 iters/sec. Estimated time to finish: 162 days, 3:44:23.025049.                                                         
 ^[[4A^[[J/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/nn/parallel/_functions.py:61: UserWarning: Was  asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.                   warnings.warn('Was asked to gather along dimension 0, but all '                                                                 
 Exception in main training loop: CUDA out of memory. Tried to allocate 372.62 MiB (GPU 0; 22.38 GiB total capacity; 21.07 GiB      already allocated; 354.00 MiB free; 230.84 MiB cached)                                                                             Traceback (most recent call last):                                                                                                
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/trainer.py", line 316, in run        
     update()                                                                                                                      
   File "/mnt/cephfs2/asr/users/fanlu/espnet/espnet/asr/pytorch_backend/asr.py", line 190, in update                               
     self.update_core()                                                                                                            
   File "/mnt/cephfs2/asr/users/fanlu/espnet/espnet/asr/pytorch_backend/asr.py", line 167, in update_core                          
     loss.backward()  # Backprop                                                                                                   
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/tensor.py", line 102, in backward               
     torch.autograd.backward(self, gradient, retain_graph, create_graph)                                                           
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/autograd/__init__.py", line 90, in backward     
     allow_unreachable=True)  # allow_unreachable flag                                                                             
 Will finalize trainer extensions and updater before reraising the exception.                                                      
 Traceback (most recent call last):                                                                                                
   File "/mnt/cephfs2/asr/users/fanlu/espnet/egs/aishell/asr3.1/../../../espnet/bin/asr_train.py", line 382, in <module>           
     main(sys.argv[1:])                                                                                                            
   File "/mnt/cephfs2/asr/users/fanlu/espnet/egs/aishell/asr3.1/../../../espnet/bin/asr_train.py", line 370, in main               
     train(args)                                                                                                                   
   File "/mnt/cephfs2/asr/users/fanlu/espnet/espnet/asr/pytorch_backend/asr.py", line 550, in train                                
     trainer.run()                                                                                                                 
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/trainer.py", line 349, in run        
     six.reraise(*exc_info)                                                                                                        
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/six.py", line 693, in reraise                         
     raise value                                                                                                                   
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/chainer/training/trainer.py", line 316, in run        
     update()                                                                                                                      
   File "/mnt/cephfs2/asr/users/fanlu/espnet/espnet/asr/pytorch_backend/asr.py", line 190, in update                               
     self.update_core()                                                                                                            
   File "/mnt/cephfs2/asr/users/fanlu/espnet/espnet/asr/pytorch_backend/asr.py", line 167, in update_core                          
     loss.backward()  # Backprop                                                                                                   
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/tensor.py", line 102, in backward               
     torch.autograd.backward(self, gradient, retain_graph, create_graph)                                                           
   File "/mnt/cephfs2/asr/users/fanlu/miniconda3/lib/python3.7/site-packages/torch/autograd/__init__.py", line 90, in backward     
     allow_unreachable=True)  # allow_unreachable flag                                                                             
 RuntimeError: CUDA out of memory. Tried to allocate 372.62 MiB (GPU 0; 22.38 GiB total capacity; 21.07 GiB already allocated; 354. 00 MiB free; 230.84 MiB cached)
```

**Comment by fanlu** [comment]
Another question.
Is the ppl result below strange?
```
epoch       iteration   perplexity  val_perplexity  elapsed_time
0           100         1.00007                     87.8458
0           200         1.00007                     158.037
0           300         1.00006                     232.062
0           400         1.00007                     306.592
0           500         1.00006                     384.287
0           600         1.00006                     453.27
0           700         1.00005                     528.082
0           800         1.00005                     601.221
0           900         1.00006                     667.406
0           1000        1.00005                     744.973
0           1100        1.00005                     1084.94
     total [..................................................]  0.86%
this epoch [########..........................................] 17.29%
      1100 iter, 0 epoch / 20 epochs
    1.0028 iters/sec. Estimated time to finish: 1 day, 10:56:33.393035.
```

**Comment by sw005320** [comment]
This looks strange. I recommend you to check the original data (or there may be a bug in espnet).

**Comment by fanlu** [comment]
> This looks strange. I recommend you to check the original data (or there may be a bug in espnet).

Yes. I run tedlium3 recipe from stage -1, data download. I suspect that is a bug in this pr

**Comment by ShigekiKarita** [comment]
I think its bug is here
https://github.com/espnet/espnet/blob/8de107f69035e0db29db51ea615e6db8e1c7b2b6/espnet/lm/pytorch_backend/lm.py#L66
It should be `torch.nn.CrossEntropy(reduction="sum")` because its value will be normalized by `count`
https://github.com/espnet/espnet/blob/8de107f69035e0db29db51ea615e6db8e1c7b2b6/espnet/lm/pytorch_backend/lm.py#L313
Therefore, currently, the reported loss is double normalized by pytorch and espnet.

**Pull request merged**